File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/OpenGLBorrowedTextureDescriptor.java
package org.maplibre.nativeffi.render;

import java.util.Objects;

/** Mutable descriptor for OpenGL caller-owned texture render targets. */
public final class OpenGLBorrowedTextureDescriptor {
  private RenderTargetExtent extent = new RenderTargetExtent();
  private OpenGLContextDescriptor context = new WglContextDescriptor();
  private int texture;
  private int target;

  public RenderTargetExtent extent() {
    return extent;
  }

  public OpenGLBorrowedTextureDescriptor extent(RenderTargetExtent extent) {
    this.extent = Objects.requireNonNull(extent, "extent");
    return this;
  }

  public OpenGLContextDescriptor context() {
    return context;
  }

  public OpenGLBorrowedTextureDescriptor context(OpenGLContextDescriptor context) {
    this.context = Objects.requireNonNull(context, "context");
    return this;
  }

  public int texture() {
    return texture;
  }

  public OpenGLBorrowedTextureDescriptor texture(int texture) {
    this.texture = texture;
    return this;
  }

  public int target() {
    return target;
  }

  public OpenGLBorrowedTextureDescriptor target(int target) {
    this.target = target;
    return this;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/OpenGLBorrowedTextureDescriptor.kt
package org.maplibre.nativeffi.render

/** Mutable descriptor for OpenGL caller-owned texture render targets. */
public class OpenGLBorrowedTextureDescriptor(
  extent: RenderTargetExtent = RenderTargetExtent(),
  context: OpenGLContextDescriptor = WglContextDescriptor(),
  texture: Int = 0,
  target: Int = 0,
) {
  public var extent: RenderTargetExtent = extent
    private set

  public var context: OpenGLContextDescriptor = context
    private set

  public var texture: Int = texture
    private set

  public var target: Int = target
    private set

  public fun extent(extent: RenderTargetExtent): OpenGLBorrowedTextureDescriptor = apply {
    this.extent = extent
  }

  public fun context(context: OpenGLContextDescriptor): OpenGLBorrowedTextureDescriptor = apply {
    this.context = context
  }

  public fun texture(texture: Int): OpenGLBorrowedTextureDescriptor = apply {
    this.texture = texture
  }

  public fun target(target: Int): OpenGLBorrowedTextureDescriptor = apply { this.target = target }
}
